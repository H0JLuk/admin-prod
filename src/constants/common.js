export const showCount = { formatter: ({ count, maxLength }) => `Осталось символов ${maxLength - count}` };

/* '^(https?:\\/\\/)?'+ // protocol
    '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|'+ // domain name
    '((\\d{1,3}\\.){3}\\d{1,3}))'+ // OR ip (v4) address
    '(\\:\\d+)?(\\/[-a-z\\d%_.~+{}]*)*'+ // port and path
    '(\\?[;&a-z\\d%_.~+=-{}]*)?'+ // query string
    '(\\#[-a-z\\d_]*)?$','i'); // fragment locator */
export const URL_REGEXP = /^(https?:\/\/)?((([a-z\d]([a-z\d-]*[a-z\d])*)\.)+[a-z]{2,}|((\d{1,3}\.){3}\d{1,3}))(:\d+)?(\/[-A-Za-z\d%_.~+#]*)*(\?[;&a-z\d%_.~+=-{}]*)?(#[-a-z\d_]*)?$/;
