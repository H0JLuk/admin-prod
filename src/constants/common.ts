import { TextAreaProps } from 'antd/lib/input';

export const showCount: TextAreaProps['showCount'] = {
    formatter: ({ count, maxLength }) => `Осталось символов ${maxLength! - count}`,
};

export const DEFAULT_SLEEP_TIME = 500; // in milliseconds

export enum DIRECTION {
    ASC = 'ASC',
    DESC = 'DESC',
}
