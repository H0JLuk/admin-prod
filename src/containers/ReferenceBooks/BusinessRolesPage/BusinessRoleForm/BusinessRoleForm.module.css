.container {
    width: 100%;
    padding: 0px 60px 0px 30px;
}

.formWrapper {
    padding: 32px;
    background-color: #f8f8f8;
}

.formContainer {
    max-width: 800px;
}

.title {
    font-weight: 600;
    font-size: 20px;
    line-height: 28px;
    letter-spacing: -.38px;
    margin-top: 24px;
    margin-bottom: 16px;
}

.datePicker {
    width: 100%;
    max-width: 250px;
}

.error {
    margin-top: 5px;
    color: #dd283b;
}

.btnGroup {
    margin: 24px 0;
}

.btnGroup > button {
    margin-right: 12px;
}

.noEditField {
    display: flex;
    align-items: center;
    color: #494949;
    font-size: 14px;
    font-weight: bold;
    min-height: 32px;
    max-width: 300px;
}
