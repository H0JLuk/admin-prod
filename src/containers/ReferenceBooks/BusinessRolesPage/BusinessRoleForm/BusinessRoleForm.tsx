import React, { useEffect, useRef, useState } from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { StaticContext } from 'react-router';
import { Col, Form, Row, Input, Button } from 'antd';
import Loading from '@components/Loading';
import {
    createBusinessRole,
    deleteBusinessRole,
    editBusinessRole,
    getBusinessRoleById,
} from '@apiServices/businessRoleService';
import { customNotifications } from '@utils/notifications';
import { confirmModal } from '@utils/utils';
import { getFormattedDate } from '@utils/helper';
import { BUTTON_TEXT } from '@constants/common';
import { BusinessRoleDto, SaveBusinessRoleRequest } from '@types';
import { FORM_RULES, getPatternAndMessage } from '@utils/validators';

import styles from './BusinessRoleForm.module.css';

type LocationState = {
    businessRole?: BusinessRoleDto;
};

export type BusinessRoleFormRouteProps = RouteComponentProps<{ businessRoleId: string; }, StaticContext, LocationState>;

interface BusinessRoleFormProps extends BusinessRoleFormRouteProps {
    matchPath: string;
    mode: string;
}

type FormDataBusinessRole = SaveBusinessRoleRequest & {
    startDate?: string;
    endDate?: string;
};

const NEW_BUSINESS_ROLE_TITLE = 'Новая бизнес-роль';

export const PLACEHOLDER = {
    NAME: 'Укажите название',
    DESCRIPTION: 'Укажите описание',
};

const LABEL = {
    NAME: 'Название',
    DESCRIPTION: 'Описание',
    START_DATE: 'Начало действия',
};

const ACTION_TEXT_BY_TYPE = {
    create: 'создана',
    edit: 'обновлена',
    delete: 'удалена',
};

function getNotifyMessage(type: keyof typeof ACTION_TEXT_BY_TYPE, name: string) {
    return <span>Бизнес-роль <b>{name}</b> успешно {ACTION_TEXT_BY_TYPE[type]}</span>;
}

function getInitialValue({ name, description, startDate }: BusinessRoleDto) {
    return {
        name: name ?? '',
        description: description ?? '',
        startDate: startDate as string,
    };
}

const BusinessRoleForm: React.FC<BusinessRoleFormProps> = ({
    matchPath,
    mode,
    location,
    history,
    match,
}) => {
    const [form] = Form.useForm();
    const [loading, setLoading] = useState(true);
    const [isSubmitDisabled, setIsSubmitDisabled] = useState(true);
    const [error, setError] = useState('');
    const { businessRoleId } = match.params;
    const initialData = useRef({} as FormDataBusinessRole);
    const isEdit = mode === 'edit';

    const goToBusinessRolesList = (redirect?: boolean) => history[redirect ? 'replace' : 'push'](matchPath);

    useEffect(() => {
        (async () => {
            if (isEdit) {
                if (!businessRoleId) {
                    goToBusinessRolesList(true);
                    return;
                }

                let { businessRole } = location.state || {};

                if (!businessRole) {
                    businessRole = await getBusinessRoleById(businessRoleId);

                    if (!businessRole) {
                        goToBusinessRolesList(true);
                        return;
                    }

                    businessRole = {
                        ...businessRole,
                        startDate: getFormattedDate(businessRole.startDate),
                        endDate: getFormattedDate(businessRole.endDate),
                    };

                    history.replace(location.pathname, { businessRole });
                }

                initialData.current = getInitialValue(businessRole);
                form.setFieldsValue(initialData.current);
            }

            setLoading(false);
        })();
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const onFinish = async (formData: SaveBusinessRoleRequest) => {
        setLoading(true);
        try {
            if (isEdit) {
                await editBusinessRole(businessRoleId, formData);
                history.replace(location.pathname);
            } else {
                await createBusinessRole(formData);
            }

            customNotifications.success({
                message: getNotifyMessage(isEdit ? 'edit' : 'create', formData.name),
            });
            goToBusinessRolesList();
        } catch (e) {
            setError(e.message);
            setLoading(false);
        }
    };

    const deleteRole = async () => {
        setLoading(true);
        try {
            await deleteBusinessRole(businessRoleId);
            customNotifications.success({
                message: getNotifyMessage('delete', initialData.current.name),
            });
            goToBusinessRolesList(true);
        } catch (e) {
            setLoading(false);
            setError(e.message);
        }
    };

    const onDeleteClick = () => {
        confirmModal({
            onOk: deleteRole,
            title: <span>Вы действительно хотите удалить бизнес-роль <b>{initialData.current.name}</b></span>,
        });
    };

    return (
        <>
            {loading && <Loading />}
            <div className={styles.container}>
                <div className={styles.title}>
                    {isEdit ? initialData.current.name : NEW_BUSINESS_ROLE_TITLE}
                </div>
                <div className={styles.formWrapper}>
                    <Form
                        id="businessRole"
                        className={styles.formContainer}
                        form={form}
                        layout="vertical"
                        onFinish={onFinish}
                        onFieldsChange={isEdit ? (() => setIsSubmitDisabled(false)) : undefined}
                        validateTrigger="onSubmit"
                    >
                        <Row gutter={[24, 6]}>
                            <Col span={12}>
                                <Form.Item
                                    name="name"
                                    label={LABEL.NAME}
                                    rules={[
                                        FORM_RULES.REQUIRED,
                                        {
                                            ...getPatternAndMessage('businessRole', 'name'),
                                        },
                                    ]}
                                    required
                                >
                                    <Input
                                        placeholder={PLACEHOLDER.NAME}
                                        allowClear
                                        maxLength={50}
                                    />
                                </Form.Item>
                            </Col>

                            <Col span={12}>
                                <Form.Item
                                    name="description"
                                    label={LABEL.DESCRIPTION}
                                    rules={[
                                        {
                                            ...getPatternAndMessage('businessRole', 'description'),
                                        },
                                    ]}
                                >
                                    <Input.TextArea
                                        placeholder={PLACEHOLDER.DESCRIPTION}
                                        rows={2}
                                        maxLength={300}
                                    />
                                </Form.Item>
                            </Col>

                            {isEdit && (
                                <Col span={12}>
                                    <div>{LABEL.START_DATE}</div>
                                    <div className={styles.noEditField}>
                                        {form.getFieldValue('startDate')}
                                    </div>
                                </Col>
                            )}
                        </Row>
                    </Form>
                    {error && <span className={styles.error}>{error}</span>}
                </div>
                <div className={styles.btnGroup}>
                    <Button onClick={() => goToBusinessRolesList()}>
                        {BUTTON_TEXT.CANCEL}
                    </Button>
                    <Button
                        htmlType="submit"
                        form="businessRole"
                        type="primary"
                        disabled={isEdit && isSubmitDisabled}
                    >
                        {isEdit ? BUTTON_TEXT.SAVE : BUTTON_TEXT.ADD}
                    </Button>
                    {isEdit && (
                        <Button
                            type="primary"
                            onClick={onDeleteClick}
                            danger
                        >
                            {BUTTON_TEXT.DELETE}
                        </Button>
                    )}
                </div>
            </div>
        </>
    );
};

export default BusinessRoleForm;
