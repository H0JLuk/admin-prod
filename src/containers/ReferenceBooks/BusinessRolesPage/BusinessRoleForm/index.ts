import BusinessRoleForm from './BusinessRoleForm';

export default BusinessRoleForm;
