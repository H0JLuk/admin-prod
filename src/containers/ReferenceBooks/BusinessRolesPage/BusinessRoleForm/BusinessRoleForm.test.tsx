import React from 'react';
import { fireEvent, render, act, screen, waitFor } from '@testing-library/react';
import BusinessRoleForm, { PLACEHOLDER } from './BusinessRoleForm';
import {
    createBusinessRole,
    deleteBusinessRole,
    editBusinessRole,
    getBusinessRoleById,
} from '@apiServices/businessRoleService';
import { sleep } from '@setupTests';
import { BUTTON_TEXT } from '@constants/common';
import { testBusinessRole } from '@testConstants';

jest.mock('@apiServices/businessRoleService', () => ({
    createBusinessRole: jest.fn(),
    deleteBusinessRole: jest.fn(),
    editBusinessRole: jest.fn(),
    getBusinessRoleById: jest.fn(),
}));

const TEST = 'Тест';

const mockHistoryPush = jest.fn();
const mockHistoryReplace = jest.fn();

const match = (id: string | null = '1') => ({
    params: {
        businessRoleId: id,
    },
});

const BusinessRolesListTestProps: any = {
    matchPath: 'test',
    mode: 'add',
    location: {
        state: testBusinessRole,
        pathname: 'test',
    },
    history: {
        push: mockHistoryPush,
        replace: mockHistoryReplace,
    },
};

const errorMessage = 'error';

describe('BusinessRoleList test', () => {
    const Component = (mode = 'add', matchParams = match()) => render(<BusinessRoleForm {...BusinessRolesListTestProps} mode={mode} match={matchParams} />);

    beforeEach(() => {
        (createBusinessRole as jest.Mock).mockResolvedValue({ id: 1, message: 'Successful', status: 'Ok' });
        (deleteBusinessRole as jest.Mock).mockResolvedValue({ message: 'Successful', status: 'Ok' });
        (editBusinessRole as jest.Mock).mockResolvedValue({ message: 'Successful', status: 'Ok' });
        (getBusinessRoleById as jest.Mock).mockResolvedValue(testBusinessRole);
    });

    it('should match snapshot', async () => {
        const RenderComponent = Component();
        await sleep();
        expect(RenderComponent.container).toMatchSnapshot();
    });

    it('should add business role', async () => {
        await waitFor(() => {
            Component();
        });

        await act(async () => {
            fireEvent.change(screen.getByPlaceholderText(PLACEHOLDER.NAME), { target: { value: TEST } });
        });

        await act(async () => {
            fireEvent.change(screen.getByPlaceholderText(PLACEHOLDER.DESCRIPTION), { target: { value: TEST } });
        });

        await act(async () => {
            fireEvent.submit(document.querySelector('#businessRole') as Element);
        });

        expect(createBusinessRole).toBeCalledTimes(1);
    });

    it('should add business role with error', async () => {
        (createBusinessRole as jest.Mock).mockRejectedValue({ message: errorMessage });

        await waitFor(() => {
            Component();
        });

        await act(async () => {
            fireEvent.change(screen.getByPlaceholderText(PLACEHOLDER.NAME), { target: { value: TEST } });
        });

        await act(async () => {
            fireEvent.change(screen.getByPlaceholderText(PLACEHOLDER.DESCRIPTION), { target: { value: TEST } });
        });

        await act(async () => {
            fireEvent.submit(document.querySelector('#businessRole') as Element);
        });

        expect(createBusinessRole).toBeCalledTimes(1);
        expect(screen.getByText(errorMessage)).toBeTruthy();
    });

    it('should edit business role', async () => {
        await waitFor(() => {
            Component('edit');
        });

        await act(async () => {
            fireEvent.change(screen.getByPlaceholderText(PLACEHOLDER.NAME), { target: { value: 'Тестовое имя.' } });
        });

        await act(async () => {
            fireEvent.change(screen.getByPlaceholderText(PLACEHOLDER.DESCRIPTION), { target: { value: TEST } });
        });

        await act(async () => {
            fireEvent.submit(document.querySelector('#businessRole') as Element);
        });

        expect(editBusinessRole).toBeCalledTimes(1);
    });

    it('should delete business role', async () => {
        await waitFor(() => {
            Component('edit');
        });

        await act(async () => {
            fireEvent.click(screen.getByText(BUTTON_TEXT.DELETE));
        });

        await sleep();

        await act(async () => {
            fireEvent.click(screen.getByText(/Да/));
        });

        expect(deleteBusinessRole).toBeCalledTimes(1);
    });

    it('should call error delete business role', async () => {
        (deleteBusinessRole as jest.Mock).mockRejectedValue({ message: errorMessage });

        await waitFor(() => {
            Component('edit');
        });

        await act(async () => {
            fireEvent.click(screen.getByText(BUTTON_TEXT.DELETE));
        });

        await sleep();

        await act(async () => {
            fireEvent.click(screen.getByText(/Да/));
        });

        expect(screen.getByText(errorMessage)).toBeTruthy();
    });

    it('should redirect to business role list', async () => {
        await waitFor(() => {
            Component('edit');
        });

        await act(async () => {
            fireEvent.click(screen.getByText(BUTTON_TEXT.CANCEL));
        });

        expect(mockHistoryPush).toBeCalledTimes(1);
    });

    it('should not find business role id', async () => {
        await waitFor(() => {
            Component('edit', match(null));
        });

        expect(mockHistoryReplace).toBeCalledTimes(1);
    });

    it('should return error for business role id', async () => {
        (getBusinessRoleById as jest.Mock).mockResolvedValue('');

        await waitFor(() => {
            Component('edit', match());
        });

        expect(mockHistoryReplace).toBeCalledTimes(1);
    });

});
