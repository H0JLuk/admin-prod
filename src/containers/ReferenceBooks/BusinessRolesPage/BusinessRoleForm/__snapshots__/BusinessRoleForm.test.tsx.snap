// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusinessRoleList test should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="title"
    >
      Новая бизнес-роль
    </div>
    <div
      class="formWrapper"
    >
      <form
        class="ant-form ant-form-vertical formContainer"
        id="businessRole"
      >
        <div
          class="ant-row"
          style="margin: -3px -12px -3px -12px;"
        >
          <div
            class="ant-col ant-col-12"
            style="padding: 3px 12px 3px 12px;"
          >
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-form-item-label"
              >
                <label
                  class="ant-form-item-required"
                  for="name"
                  title="Название"
                >
                  Название
                </label>
              </div>
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <span
                      class="ant-input-affix-wrapper"
                    >
                      <input
                        class="ant-input"
                        id="name"
                        maxlength="50"
                        placeholder="Укажите название"
                        type="text"
                        value=""
                      />
                      <span
                        class="ant-input-suffix"
                      >
                        <span
                          aria-label="close-circle"
                          class="anticon anticon-close-circle ant-input-clear-icon-hidden ant-input-clear-icon"
                          role="button"
                          tabindex="-1"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="close-circle"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                            />
                          </svg>
                        </span>
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-col ant-col-12"
            style="padding: 3px 12px 3px 12px;"
          >
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-form-item-label"
              >
                <label
                  class=""
                  for="description"
                  title="Описание"
                >
                  Описание
                </label>
              </div>
              <div
                class="ant-col ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <textarea
                      class="ant-input"
                      id="description"
                      placeholder="Укажите описание"
                      rows="2"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
      
    </div>
    <div
      class="btnGroup"
    >
      <button
        class="ant-btn"
        type="button"
      >
        <span>
          Отменить
        </span>
      </button>
      <button
        class="ant-btn ant-btn-primary"
        form="businessRole"
        type="submit"
      >
        <span>
          Добавить
        </span>
      </button>
    </div>
  </div>
</div>
`;
